### PR TITLE
Promisify RemoteReporter and UDPSender

### PR DESCRIPTION
**Pull request by MarckK**
Signed-off-by: Kara de la Marck <karadelamarck@gmail.com>

### CODE REVIEW

File: src/reporters/remote_reporter.js [hunk 1]
```diff
@@ -45,29 +45,32 @@ export default class RemoteReporter {
...
   report(span: Span): void {
-    let response: SenderResponse = this._sender.append(ThriftUtils.spanToThrift(span));
-    if (response.err) {
-      this._logger.error('Failed to append spans in reporter.');
-      this._metrics.reporterDropped.increment(response.numSpans);
-    }
+    let responsePromise: Promise<SenderResponse> = this._sender.append(ThriftUtils.spanToThrift(span));
+
+    responsePromise.then(response => {
+      this._sender.append(ThriftUtils.spanToThrift(span));
```
Comment (yurishkuro) [inline src/reporters/remote_reporter.js]: is this needed?

File: src/reporters/udp_sender.js [hunk 2]
```diff
@@ -148,7 +148,7 @@ export default class UDPSender {
     });
     this._reset();
```
Comment (yurishkuro) [inline src/reporters/udp_sender.js]: I think reset needs to be called at L134, after the spans are converted to thrift buffer

File: src/reporters/udp_sender.js [hunk 3]
```diff
@@ -148,7 +148,7 @@ export default class UDPSender {
...
 
-    return { err: false, numSpans: numSpans };
+    return Promise.resolve({ err: false, numSpans: numSpans });
```
Comment (yurishkuro) [inline src/reporters/udp_sender.js]: is there a Promise-based version of `this._client.send()`? If not, the promise needs to be created before calling `send()` and completed in the callback at L143

File: src/_flow/reporter.js [hunk 4]
```diff
@@ -20,8 +20,8 @@ declare interface Reporter {
 }
 
 declare class Sender {
-  append(span: Span): SenderResponse;
-  flush(): SenderResponse;
+  append(span: Span): Promise<SenderResponse>;
+  flush(): Promise<SenderResponse>;
```
Comment (yurishkuro) [inline src/_flow/reporter.js]: how will the Promise class get translated to Node 0.10? Is there some backfill implementation that babel provides?

File: src/_flow/reporter.js [hunk 5]
```diff
@@ -20,8 +20,8 @@ declare interface Reporter {
...
   close(): void;
```
Comment (yurishkuro) [inline src/_flow/reporter.js]: I think close() needs to be changed to use Promise as well

**Comment by yurishkuro** [comment]
I think tracer.close() should also be changed to use promise from reporter.close()

Overall the change looks great.

File: src/_flow/reporter.js (hunk 4 above)
Comment (MarckK) [reply src/_flow/reporter.js]: There is a babel polyfill.  I'm checking if Babel will implement it for us for Node 0.10 and 0.12. 🤞

Comment (MarckK) [reply src/_flow/reporter.js]: 😌 I have so much appreciation for Babel.

File: src/reporters/udp_sender.js (hunk 2 above)
Comment (MarckK) [reply src/reporters/udp_sender.js]: Would this then close #214?

File: src/reporters/remote_reporter.js [hunk 6]
```diff
@@ -45,29 +45,30 @@ export default class RemoteReporter {
   }
 
   report(span: Span): void {
-    let response: SenderResponse = this._sender.append(ThriftUtils.spanToThrift(span));
-    if (response.err) {
-      this._logger.error('Failed to append spans in reporter.');
-      this._metrics.reporterDropped.increment(response.numSpans);
-    }
+    this._sender.append(ThriftUtils.spanToThrift(span)).then(response => {
+      this._sender.append(ThriftUtils.spanToThrift(span));
```
Comment (yurishkuro) [inline src/reporters/remote_reporter.js]: duplicate

File: src/reporters/composite_reporter.js [hunk 7]
```diff
@@ -30,23 +30,9 @@ export default class CompositeReporter {
...
-  }
-
-  close(callback: ?Function): void {
-    let modifiedCallback: ?Function = this.compositeCallback(callback);
-    this._reporters.forEach(r => {
-      r.close(modifiedCallback);
-    });
+  close(): Promise<void> {
+    let promises = this._reporters.map(r => r.close());
+    return Promise.all(promises).then(() => {});
```
[the remark this answers is not in the document]
Comment (MarckK) [reply src/reporters/composite_reporter.js]: Unfortunately, generates flow errors. Might be better to take out the option to pass callbacks into RemoteReporter close method.

**Pull request closed without merging**